### PR TITLE
Fix: StringArray non-extensible due to inconsisent assertion

### LINKED ISSUE #34309: BUG: StringArray non-extensible due to inconsisent assertion
- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the latest version of pandas.

- [X] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
import pandas as pd
from pandas import StringDtype
from pandas.core.arrays import StringArray
from pandas.core.dtypes.dtypes import register_extension_dtype

@register_extension_dtype
class MyExtensionDtype(StringDtype):
    name = 'my_extension'

    def __repr__(self) -> str:
        return "MyExtensionDtype"

    @classmethod
    def construct_array_type(cls) -> "Type[MyExtensionStringArray]":
        return MyExtensionStringArray

class MyExtensionStringArray(StringArray):
    def __init__(self, values, copy=False):
        super().__init__(values, copy)
        self._dtype = MyExtensionDtype()

series = pd.Series(["test", "test2"], dtype="my_extension")
assert series.dtype == 'my_extension'
```

Results in
`assert dtype == "string" AssertionError`

#### Problem description

It should be possible to extend the StringDtype/StringArray for users to design efficient subtypes. I believe that the the AssertionError is a bug and not intended, as pandas wants to have extensible dtypes, because there is the ExtensionDtype.

#### Expected Output
The code above should pass without errors.

PR with fix on it's way.

#### Output of ``pd.show_versions()``

<details>

pandas v1.0.3

</details>

### PR DESCRIPTION
**Pull request by sbrugman**
Fix for #34309

- [X] closes #34309
- [X] tests added / passed
- [X] passes `black pandas`
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

### CODE REVIEW

File: pandas/tests/arrays/string_/test_string.py [hunk 1]
```diff
@@ -332,3 +333,29 @@ def test_memory_usage():
     series = pd.Series(["a", "b", "c"], dtype="string")
 
     assert 0 < series.nbytes <= series.memory_usage() < series.memory_usage(deep=True)
+
+
+def test_extension_assert():
+    import pandas as pd
+    from pandas import StringDtype
+    from pandas.core.arrays import StringArray
+    from pandas.core.dtypes.dtypes import register_extension_dtype
```
Comment (dsaxton) [inline pandas/tests/arrays/string_/test_string.py:342]: Can you put these imports at the top (pandas is already imported)? Also may want to rename the test to something specific to the functionality being tested (i.e., subclassing).

File: pandas/tests/arrays/string_/test_string.py [hunk 2]
```diff
@@ -332,3 +333,29 @@ def test_memory_usage():
...
+        def construct_array_type(cls) -> "Type[MyExtensionStringArray]":
+            return MyExtensionStringArray
+
+    class MyExtensionStringArray(StringArray):
+        def __init__(self, values, copy=False):
+            super().__init__(values, copy)
+            self._dtype = MyExtensionDtype()
+
+    series = pd.Series(["test", "test2"], dtype="my_extension")
+    assert series.dtype == "my_extension"
```
Comment (dsaxton) [inline pandas/tests/arrays/string_/test_string.py:361]: Should probably also assert that the values are as expected.

**Comment by sbrugman** [comment]
@dsaxton Thanks for your review. I've processed your proposed changes and added an additional string dtype test.

File: pandas/tests/arrays/string_/test_string.py [hunk 3]
```diff
@@ -332,3 +338,26 @@ def test_memory_usage():
     series = pd.Series(["a", "b", "c"], dtype="string")
 
     assert 0 < series.nbytes <= series.memory_usage() < series.memory_usage(deep=True)
+
+
+def test_string_dtype_subclassing():
+    @register_extension_dtype
+    class MyExtensionDtype(StringDtype):
```
Comment (dsaxton) [inline pandas/tests/arrays/string_/test_string.py:345]: Not a big deal, but could also use pd.StringDtype here and below and then not have to import it above

**Review by dsaxton — APPROVED** [review]
Tiny nit, but otherwise LGTM

**Comment by sbrugman** [comment]
> Tiny nit, but otherwise LGTM

Done

**Comment by jreback** [comment]
cc @jorisvandenbossche @TomAugspurger

**Comment by TomAugspurger** [comment]
I think the issue has an ongoing discussion that needs to be resolved.

**Comment by jbrockmendel** [comment]
@sbrugman can you merge master just to keep this from drifting

File: pandas/core/arrays/string_.py [hunk 4]
```diff
@@ -196,7 +196,7 @@ def _validate(self):
     @classmethod
     def _from_sequence(cls, scalars, dtype=None, copy=False):
         if dtype:
-            assert dtype == "string"
+            assert isinstance(dtype, StringDtype)
```
[the remark this answers is not in the document]
Comment (TomAugspurger) [reply pandas/core/arrays/string_.py:199]: They're equivalent right now, but that may be changing in the future with an arrow-backed string array.

But my main objection to this was the original motivation from the issue not being a supported use case.

[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/arrays/string_.py:199]: @TomAugspurger if you dont think this should be supported, should this PR be closed?

**Pull request closed without merging**